napari/layers/utils/text_manager.py

```diff
@@ -51,92 +49,97 @@ class TextManager(EventedModel):
+
+    Private Attributes
+    ------------------
+    _n_text : int
+        The number of text elements managed, which should correspond to the number
+        of rows in the properties table. Should be removed if/when the properties
+        table reliably stores that count.
     """
 
-    values: Array[str] = []
+    # Declare properties as a generic dict so that a copy is not made on validation
+    # and we can rely on a layer and this sharing the same instance.
+    properties: dict
     visible: bool = True
     size: PositiveInt = 12
-    color: Array[float, (4,)] = 'cyan'
     blending: Blending = Blending.TRANSLUCENT
     anchor: Anchor = Anchor.CENTER
     # Use a scalar default translation to broadcast to any dimensionality.
     translation: Array[float] = 0
     rotation: float = 0
-    _mode: TextMode = TextMode.NONE
-    _text_format_string: str = ''
+    text: Union[STRING_ENCODINGS] = FormatStringEncoding(format_string='')
+    color: Union[COLOR_ENCODINGS] = ConstantColorEncoding(
+        constant=DEFAULT_COLOR
+    )
+    _n_text: int
 
-    def __init__(self, text=None, n_text=None, properties=None, **kwargs):
+    def __init__(self, *, n_text, **kwargs):
         super().__init__(**kwargs)
-        if 'values' in kwargs:
-            text = kwargs['values']
-            n_text = len(text)
-        self._set_text(text, n_text, properties=properties)
-
-    def _set_text(
-        self,
-        text: Optional[str],
-        n_text: int,
-        properties: Optional[dict] = None,
-    ):
-        if properties is None:
-            properties = {}
-        if text is None:
-            text = np.empty(0)
-        if len(properties) == 0 or len(text) == 0:
-            # set text mode to NONE if no props/text are provided
-            self._mode = TextMode.NONE
-            self._text_format_string = ''
-            self.values = np.empty(0)
-        else:
-            formatted_text, text_mode = format_text_properties(
-                text, n_text, properties
-            )
-            self._text_format_string = text
-            self._mode = text_mode
-            self.values = formatted_text
+        # Add a custom event that is emitted when text needs to be re-rendered.
+        # This means external clients do not need to reconnect to the events of
+        # any mutable fields when their instance changes.
+        self.events.add(text_update=Event)
+        # Connect to lambda's that wrap these bound methods because our Python 3.7
+        # environment cannot hash these bound methods because TextManager itself is
+        # not hashable, but newer environments can (for reasons not understood).
+        self.events.properties.connect(lambda e: self._on_properties_changed())
+        self.events.text.connect(lambda e: self._on_text_changed())
+        self.events.color.connect(lambda e: self._on_color_changed())
+        self._n_text = n_text
+        self._on_text_changed()
+        self._on_color_changed()
```

Comment:
Hmm i don't love this internal event connection and routing. Would we need any of the `_on_text_changed` etc methods >=3.8? Would be nice not to have them. Or is there any overwriting the setters that we could do. I think we did stuff like that for our layerlist stuff. Could come later, but just wanted to flag. Definitely not a blocker

Reply:
I'm not sure if this case is affected, but I just remembered this issue that @Czaki raised: #3217 (and its fix #3217). If this is a problem here, then the lambdas should be replaced!

Reply:
> Yes, please use `napari.utils.events.event_utils.connect_no_arg` for this.

Thanks for the pointer to `connect_no_arg` - that worked in my case too!

I understand the need for that in terms of lifetimes of objects - we don't want to keep an object around just because an event emitter has a reference to a method bound to it. Though, the implementation of `connect_no_arg` initially seems a little dangerous to me - what happens when `ref()` in `_def cb` returns `None`? Or is that not possible?

But I still don't understand why this (or the lambda, or Python 3.8+) fixes the bug I encountered in 3.7 (i.e. a method bound to an instance of `TextManager` is not hashable because `TextManager` is not hashable) and [weakref shouldn't help with that](https://docs.python.org/3/library/weakref.html#weakref.ref). Any clarification would be much appreciated! Alternatively, we might want to consider changing `EventEmitter._blocked` to be a `List` instead of a `Dict` (removing the need for hash), though I'm not sure how easy that is.

Reply:
> Hmm i don't love this internal event connection and routing. Would we need any of the `_on_text_changed` etc methods >=3.8?

These aren't related to the Python version related bug (only the use of `lambda` or `connect_no_arg`). This is more about playing nicely with pydantic. I guess we could something like the following:

```python
    def __setattr__(self, key, value):
        super().__setattr__(key, value)
        if key == 'text':
            self._on_text_changed()
        elif key == 'color':
            self._on_color_changed()
```

instead of using connections, but I don't love referring to attribute names as strings at runtime, unless we don't have any alternative. I don't see another alternative in `layerlist.py`, but let me know if I missed it.

Reply:
> Though, the implementation of `connect_no_arg` initially seems a little dangerous to me - what happens when `ref()` in `_def cb` returns `None`? Or is that not possible?

It may end with AttributeError. I do not observe such error. Maybe it is a good idea to add `try ... except`. @tlambert03  What did you think?

Reply:
> instead of using connections, but I don't love referring to attribute names as strings at runtime, unless we don't have any alternative.

I maybe have a slight preference for this, but curious if @tlambert03 has thoughts